### PR TITLE
Avoid crash on overflow in numincrby or nummultby

### PR DESCRIPTION
**Pull request by oshadmi**
Incrementing beyond range of number is crashing redis
```
127.0.0.1:6379> JSON.SET key $ 1.6350000000001314e+308
OK
127.0.0.1:6379> JSON.NUMINCRBY key $ 1.6350000000001314e+308
```

The problem was that `update` method was taking the value being updated, e.g., then applying a closure on it, and then setting the value to the value returned from the closure, and 2 problems could have been caused:
1) The closure could have failed, e.g., on a numeric overflow - and was unwrapped
2) Value was not being restored upon error, leaving a null in place of the previous value

The fix is to handle errors properly, and to set the value by ref, so it is not lost on error.
No public API was broken.

More details:
About `replace`
```
///
/// Replaces a value at a given `path`, starting from `root`
///
/// The new value is the value returned from `func`, which is called on the current value.
///
/// If the returned value from `func` is `None`, the current value is removed.
/// If the returned value from `func` is `Err`, the current value remains (although it could be modified by `func`)
///
fn replace<F: FnMut(&mut IValue) -> Result<Option<IValue>, Error>>
```

About `update`
```
///
/// Updates a value at a given `path`, starting from `root`
///
/// The value is modified by `func`, which is called on the current value.
/// If the returned value from `func` is `None`, the current value is removed.
/// If the returned value from `func` is `Err`, the current value remains (although it could be modified by `func`)
///
fn update<F: FnMut(&mut IValue) -> Result<Option<()>, Error>>
```

### CODE REVIEW

File: tests/pytest/test.py [hunk 1]
```diff
@@ -807,6 +807,12 @@ def testNumIncrCommand(env):
     r.assertEqual(1, res['foo'])
     r.assertEqual(84, res['bar'])
 
+    # test overflow
+    r.assertOk(r.execute_command('JSON.SET', 'big_num', '.', '1.6350000000001313e+308'))
+    r.expect('JSON.NUMINCRBY', 'big_num', '.', '1.6350000000001313e+308').raiseError()
+    r.expect('JSON.NUMMULTBY', 'big_num', '.', '2').raiseError()
+    # (value remains)
+    r.assertEqual(r.execute_command('JSON.GET', 'big_num', '.'), '1.6350000000001313e308')
```
Comment (MeirShpilraien) [inline tests/pytest/test.py:815]: Can we also test it on a none root element?

**Review by MeirShpilraien — COMMENTED** [review]
Suggestion

File: src/ivalue_manager.rs [hunk 2]
```diff
@@ -46,11 +46,16 @@ fn update<F: FnMut(IValue) -> Result<Option<IValue>, Error>>(
...
-                        let v = e.insert(IValue::NULL);
-                        if let Some(res) = (func)(v)? {
-                            e.insert(res);
-                        } else {
-                            e.remove();
+                        let mut v = e.get_mut();
+                        match (func)(&mut v) {
+                            Ok(res) => {
+                                if let Some(res) = res {
+                                    *v = res;
```
Comment (MeirShpilraien) [inline src/ivalue_manager.rs:53]: Maybe I am missing something but how this API allows me to edit the existing entry and keep it?

File: src/ivalue_manager.rs [hunk 3]
```diff
@@ -63,11 +68,81 @@ fn update<F: FnMut(IValue) -> Result<Option<IValue>, Error>>(
...
-                            let v = std::mem::replace(&mut arr[x], IValue::NULL);
-                            if let Some(res) = (func)(v)? {
-                                arr[x] = res;
-                            } else {
-                                arr.remove(x);
+                            let mut v = &mut arr.as_mut_slice()[x];
+                            match (func)(&mut v) {
+                                Ok(res) => {
+                                    if let Some(res) = res {
+                                        *v = res;
```
Comment (MeirShpilraien) [inline src/ivalue_manager.rs:75]: Same question here, how do I edit existing entry and keep it?

File: src/ivalue_manager.rs [hunk 4]
```diff
@@ -63,11 +68,81 @@ fn update<F: FnMut(IValue) -> Result<Option<IValue>, Error>>(
...
+                                    }
+                                }
+                                Err(err) => return Err(err),
+                            }
+                        }
+                        return Ok(());
+                    }
+                    arr.get_mut(x)
+                } else {
+                    None
```
Comment (MeirShpilraien) [inline src/ivalue_manager.rs:87]: Can this happened? if not asset.

File: src/ivalue_manager.rs [hunk 5]
```diff
@@ -63,11 +68,81 @@ fn update<F: FnMut(IValue) -> Result<Option<IValue>, Error>>(
...
+                }
+            }
+            _ => None,
```
Comment (MeirShpilraien) [inline src/ivalue_manager.rs:90]: Can this happened? if not asset.

File: src/ivalue_manager.rs [hunk 6]
```diff
@@ -63,11 +68,81 @@ fn update<F: FnMut(IValue) -> Result<Option<IValue>, Error>>(
...
+            target = t;
+        } else {
+            break;
+        }
+    }
+
+    Ok(())
+}
+
+fn update_in_closure<F: FnMut(&mut IValue) -> Result<Option<()>, Error>>(
```
Comment (MeirShpilraien) [inline src/ivalue_manager.rs:103]: Can you add a documentation to this function that explain where is it used and how is it different from the one above?

File: src/ivalue_manager.rs [hunk 7]
```diff
@@ -247,25 +337,24 @@ impl<'a> WriteHolder<IValue, IValue> for IValueKeyHolderWrite<'a> {
                     o.insert(key.to_string(), v.take());
                 }
             }
-            let val = root.take();
-            self.set_root(Some(val))?;
         } else {
-            update(&path, self.get_value().unwrap().unwrap(), |mut val| {
-                if let Some(o) = val.as_object_mut() {
+            update_in_closure(&path, self.get_value().unwrap().unwrap(), |val| {
```
Comment (MeirShpilraien) [inline src/ivalue_manager.rs:341]: I guess this allows you to update an existing value, maybe call one update and the other replace?

File: src/ivalue_manager.rs [hunk 8]
```diff
@@ -287,9 +376,12 @@ impl<'a> WriteHolder<IValue, IValue> for IValueKeyHolderWrite<'a> {
     fn bool_toggle(&mut self, path: Vec<String>) -> Result<bool, RedisError> {
         let mut res = None;
         self.do_op(path, |v| {
-            let val = v.to_bool().unwrap() ^ true;
-            res = Some(val);
-            Ok(Some(IValue::from(val)))
+            if let DestructuredMut::Bool(mut bool_mut) = v.destructure_mut() {
```
Comment (MeirShpilraien) [inline src/ivalue_manager.rs:379]: Can you explain what `DestructuredMut` is for?

File: src/manager.rs [hunk 9]
```diff
@@ -186,8 +186,16 @@ impl<'a> KeyHolderWrite<'a> {
         if paths.is_empty() {
             // updating the root require special treatment
             let root = self.get_value().unwrap().unwrap();
-            let res = (op_fun)(root.take())?;
-            self.set_root(res)?;
+            let prev_val = root.take();
+            let rollback_val = prev_val.clone();
```
Comment (MeirShpilraien) [inline src/manager.rs:190]: So you do use clone?

File: tests/pytest/test.py [hunk 10]
```diff
@@ -645,22 +647,30 @@ def testArrInsertCommand(env):
...
+
+        r.assertEqual(r.execute_command('JSON.ARRINSERT', 'test', jpath, 1, '5'), 5)
+        r.assertEqual(r.execute_command('JSON.ARRINSERT', 'test', jpath, -2, '6'), 6)    
+        r.assertEqual(r.execute_command('JSON.GET', 'test', jpath), "[3,5,2,6,1,4]")
+        
+        r.assertEqual(r.execute_command('JSON.ARRINSERT', 'test', jpath, -3, '7', '{"A":"Z"}', '9'), 9)
+        r.assertEqual(r.execute_command('JSON.GET', 'test', jpath), '[3,5,2,7,{"A":"Z"},9,6,1,4]')
+
+        r.expect('JSON.ARRINSERT', 'test', jpath, -10, '10').raiseError()
+        if not IS_SERDE_JSON:
```
Comment (MeirShpilraien) [inline tests/pytest/test.py:666]: So the fix is not working for serde json?

File: tests/pytest/test.py [hunk 11]
```diff
@@ -645,22 +647,30 @@ def testArrInsertCommand(env):
...
+            # Value should remain
+            r.assertEqual(r.execute_command('JSON.GET', 'test', jpath), '[3,5,2,7,{"A":"Z"},9,6,1,4]')
+        r.expect('JSON.ARRINSERT', 'test', jpath, 10, '10').raiseError()
+        if not IS_SERDE_JSON:
```
Comment (MeirShpilraien) [inline tests/pytest/test.py:670]: Same comment

**Review by MeirShpilraien — COMMENTED** [review]
Looks good, added some comments. I also suggest to write the technical details of the solution on the top comment and explain why this solution is safe with regards to performance (no cloning, though I do see a place where you still clone, added a comment about that).

File: src/ivalue_manager.rs (hunk 6 above)
Comment (oshadmi) [reply src/ivalue_manager.rs:103]: Added explanation comment in `fn update`

File: src/ivalue_manager.rs (hunk 7 above)
Comment (oshadmi) [reply src/ivalue_manager.rs:341]: What about `update` and `calculate`?
(the caller is doing the actual replacement)

File: src/ivalue_manager.rs (hunk 8 above)
Comment (oshadmi) [reply src/ivalue_manager.rs:379]: It is an `Enum` used to obtain a mutable reference to `IValue`,
Specifically for Boolean, `IValue` does not directly contain Boolean,
So this is a proxy type which allows getting and setting the original `IValue` as Boolean.

File: src/manager.rs (hunk 9 above)
Comment (oshadmi) [reply src/manager.rs:190]: This is outdated.
Checking where is commit ea44cf5 ?

File: tests/pytest/test.py (hunk 10 above)
Comment (oshadmi) [reply tests/pytest/test.py:666]: Not yet done for `serde_json`
After review is done for ijson, will proceed with `serde_json`.
Do we still want to invest in it BTW?

File: src/manager.rs (hunk 9 above)
Comment (MeirShpilraien) [reply src/manager.rs:190]: How come its outdated and I see it in this PR?

File: src/ivalue_manager.rs (hunk 8 above)
Comment (MeirShpilraien) [reply src/ivalue_manager.rs:379]: Please document.

File: src/ivalue_manager.rs (hunk 4 above)
Comment (MeirShpilraien) [reply src/ivalue_manager.rs:87]: I believe that if something can not happened assert it, this way it will be easier to catch bugs.

File: src/ivalue_manager.rs (hunk 6 above)
Comment (MeirShpilraien) [reply src/ivalue_manager.rs:103]: Don't see it

File: src/ivalue_manager.rs (hunk 7 above)
Comment (MeirShpilraien) [reply src/ivalue_manager.rs:341]: How come the caller doing the replacement? The caller returns a new value and the update functions replace it no?

File: tests/pytest/test.py (hunk 10 above)
Comment (MeirShpilraien) [reply tests/pytest/test.py:666]: I believe we should still support serde json.

File: src/ivalue_manager.rs (hunk 2 above)
Comment (MeirShpilraien) [reply src/ivalue_manager.rs:53]: Lets explain this in a comment in the code and also on top comment on the PR.

File: src/ivalue_manager.rs (hunk 3 above)
Comment (MeirShpilraien) [reply src/ivalue_manager.rs:75]: Same comment as above.

File: src/manager.rs (hunk 9 above)
Comment (oshadmi) [reply src/manager.rs:190]: This was a preliminary commit which was also addressing `serde_json`.
Now we currently putting `serde_json` aside.
I suggest to keep it anyway (no need to rollback IMHO)

File: src/ivalue_manager.rs (hunk 6 above)
Comment (oshadmi) [reply src/ivalue_manager.rs:103]: Will add in code (not only in GitHub comment)

File: src/ivalue_manager.rs (hunk 7 above)
Comment (oshadmi) [reply src/ivalue_manager.rs:341]: Will rename to `update` and `replace`

File: tests/pytest/test.py (hunk 10 above)
Comment (oshadmi) [reply tests/pytest/test.py:666]: If we decide to keep maintaining `serde_json` it will be in a different PR once `ijson` is done.

**Review by MeirShpilraien — APPROVED** [review]
